source/core_mqtt.c

```diff
@@ -1024,8 +1024,11 @@ static MQTTStatus_t handleKeepAlive( MQTTContext_t * pContext )
         }
         else
         {
-            if( calculateElapsedTime( now, pContext->lastPacketTime ) > keepAliveMs )
+            if(( calculateElapsedTime( now, pContext->lastPacketTime ) > keepAliveMs ) ||
+              ( calculateElapsedTime( now, pContext->lastPacketReceivedTime) > keepAliveMs ))
             {
+                pContext->lastPacketReceivedTime = now; /* this to avoid sending back-to-back MQTT pings. Note that MQTT_Ping will
```

Comment:
Not sure this is necessary because pContext->waitingForPingResp will always be true once a ping has been sent.